prometheus/promhttp/http.go

```diff
@@ -169,15 +172,42 @@ func HandlerForTransactional(reg prometheus.TransactionalGatherer, opts HandlerO
 		header.Set(contentTypeHeader, string(contentType))
 
 		w := io.Writer(rsp)
-		if !opts.DisableCompression && gzipAccepted(req.Header) {
-			header.Set(contentEncodingHeader, "gzip")
-			gz := gzipPool.Get().(*gzip.Writer)
-			defer gzipPool.Put(gz)
+		if !opts.DisableCompression {
+			offers := defaultEncodingOffers
+			if len(opts.EncodingOffers) > 0 {
+				offers = opts.EncodingOffers
+			}
+			// TODO(mrueg): Replace internal/github.com/gddo once https://github.com/golang/go/issues/19307 is implemented.
+			compression := httputil.NegotiateContentEncoding(req, offers)
+			switch compression {
+			case "zstd":
+				header.Set(contentEncodingHeader, "zstd")
+				// TODO(mrueg): Replace klauspost/compress with stdlib implementation once https://github.com/golang/go/issues/62513 is implemented.
+				z, err := zstd.NewWriter(rsp, zstd.WithEncoderLevel(zstd.SpeedFastest))
+				if err != nil {
+					return
+				}
+
+				z.Reset(w)
+				defer z.Close()
+
+				w = z
+			case "gzip":
+				header.Set(contentEncodingHeader, "gzip")
+				gz := gzipPool.Get().(*gzip.Writer)
+				defer gzipPool.Put(gz)
+
+				gz.Reset(w)
+				defer gz.Close()
 
-			gz.Reset(w)
-			defer gz.Close()
+				w = gz
+			case "identity":
+				// This means the content is not encoded.
+			default:
+				// The content encoding was not implemented yet.
+				return
```

Comment:
What about following what I proposed in the other comment? To use opt.ErrorLog if not nil and continue without compression?

(This is for compatibility and caller error handling)

Reply:
GetWriter() throws an error that the handler catches and uses ErrorLog there.